### PR TITLE
Rework crate to be more composable

### PR DESCRIPTION
**Pull request by museun**
This a massive change, basically a rewrite and still a WIP

### CODE REVIEW

File: src/ng/messages/whisper.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    name: StrIndex,
+    data: StrIndex,
+}
+
+impl<'t> Whisper<'t> {
+    raw!();
+    tags!();
+    str_field!(name);
+    str_field!(data);
+}
```
Comment (Emilgardis) [inline src/ng/messages/whisper.rs]: missing fn user_id, badges, badge_info, color, display_name, emotes getters for tags, similar to PRIVMSG

File: src/ng/messages/cap.rs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+    pub fn acknowledged(&self) -> bool {
+        self.acknowledged
+    }
+}
+
+impl<'a> FromIrcMessage<'a> for Cap<'a> {
+    type Error = InvalidMessage;
+
+    fn from_irc(msg: IrcMessage<'a>) -> Result<Self, Self::Error> {
+        const ACK: &str = "ACK";
```
Comment (Emilgardis) [inline src/ng/messages/cap.rs]: support `NAK`?

File: src/ng/messages/clear_chat.rs [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+impl<'t> ClearChat<'t> {
+    raw!();
+    str_field!(channel);
+    opt_str_field!(name);
+    tags!();
+
+    pub fn ban_duration(&self) -> Option<u64> {
+        self.tags().get_parsed("ban-duration")
+    }
+}
```
Comment (Emilgardis) [inline src/ng/messages/clear_chat.rs]: missing fn room_id to get tag `room-id`.

File: src/ng/messages/clear_chat.rs [hunk 4]
```diff
@@ -0,0 +1,78 @@
+use crate::ng::{FromIrcMessage, InvalidMessage, Validator};
+use crate::ng::{IrcMessage, Str, StrIndex, TagIndices, Tags};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ClearChat<'t> {
+    raw: Str<'t>,
+    tags: TagIndices,
+    channel: StrIndex,
+    name: Option<StrIndex>,
```
Comment (Emilgardis) [inline src/ng/messages/clear_chat.rs]: What is this supposed to be?

afaik, there's no data in CLEARCHAT command

File: src/ng/messages/clear_msg.rs [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::ng::irc;
+
+    #[test]
+    #[cfg(feature = "serde")]
+    fn clear_msg_serde() {
+        let input = ":tmi.twitch.tv CLEARMSG #museun :HeyGuys\r\n";
```
Comment (Emilgardis) [inline src/ng/messages/clear_msg.rs]: this test should include tag `target-msg-id`

File: src/ng/messages/clear_msg.rs [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        crate::ng::serde::round_trip_json::<ClearMsg>(input);
+    }
+
+    #[test]
+    fn clear_msg() {
+        let input = ":tmi.twitch.tv CLEARMSG #museun :HeyGuys\r\n";
```
Comment (Emilgardis) [inline src/ng/messages/clear_msg.rs]: this test should include tag `target-msg-id`

File: src/ng/messages/clear_msg.rs [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        for msg in irc::parse(input).map(|s| s.unwrap()) {
+            let cm = ClearMsg::from_irc(msg).unwrap();
+            assert_eq!(cm.channel(), "#museun");
+            assert_eq!(cm.message().unwrap(), "HeyGuys");
+        }
+    }
+
+    #[test]
+    fn clear_msg_empty() {
+        let input = ":tmi.twitch.tv CLEARMSG #museun\r\n";
```
Comment (Emilgardis) [inline src/ng/messages/clear_msg.rs]: this test should include tag `target-msg-id`

File: src/ng/messages/privmsg.rs [hunk 8]
```diff
@@ -0,0 +1,251 @@
...
+            .map(parse_emotes)
+            .unwrap_or_default()
+    }
+
+    pub fn is_broadcaster(&self) -> bool {
+        self.contains_badge(BadgeKind::Broadcaster)
+    }
+
+    pub fn is_moderator(&self) -> bool {
+        self.tags().get_as_bool("mod")
```
Comment (Emilgardis) [inline src/ng/messages/privmsg.rs]: should be
```suggestion
        self.contains_badge(BadgeKind::Moderator)
```

which should be equivalent according to twitch.

File: src/ng/messages/privmsg.rs [hunk 9]
```diff
@@ -0,0 +1,251 @@
...
+pub struct Privmsg<'t> {
+    raw: Str<'t>,
+    tags: TagIndices,
+    name: StrIndex,
+    channel: StrIndex,
+    data: StrIndex,
+    ctcp: Option<StrIndex>,
+}
+
+impl<'t> Privmsg<'t> {
```
Comment (Emilgardis) [inline src/ng/messages/privmsg.rs]: missing

```rust
pub fn custom_reward_id(&self) -> Option<&str> {
    self.tags().get("custom-reward-id")
}
pub fn msg_id(&self) -> Option<&str> {
    self.tags().get("msg-id")
}
```

both from the new community points rewards, not documented in twitch irc "spec".

`msg-id` is returned on for example highlighted message, where `msg-id` would be `highlighted-message`
`custom-reward-id` is returned on custom rewards set by broadcaster, with no api from twitch to retrieve proper name, looks like a uuid.

File: src/ng/messages/user_state.rs [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+use crate::{color::Color, parse_badges, parse_emotes, Badge, BadgeInfo, Emotes};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct UserState<'t> {
+    raw: Str<'t>,
+    tags: TagIndices,
+    channel: StrIndex,
+}
+
+impl<'t> UserState<'t> {
```
Comment (Emilgardis) [inline src/ng/messages/user_state.rs]: missing user-id tag getter

**Review by Emilgardis — CHANGES_REQUESTED** [review]
nice, some minor nits, waiting for everything else but looking good so far!

File: src/ng/messages/whisper.rs (hunk 1 above)
Comment (museun) [reply src/ng/messages/whisper.rs]: Yeah, a lot of these are missing these simple getters. I have a new macro that'll even allow documentation. So hopefully when I make a 2nd pass moving the documentation over I'll grab all of the right methods.

One big question: should the serde serialization use /every/ accessor or just the fields?

File: src/ng/messages/cap.rs (hunk 2 above)
Comment (museun) [reply src/ng/messages/cap.rs]: I would assume 'not ACK' (i.e. false for `acknowledge`)  would be enough. but I can make that bool into an enum for `Acknowledge(&'a str)`, `NotAcknowledge(&'a str)`

File: src/ng/messages/clear_chat.rs (hunk 4 above)
Comment (museun) [reply src/ng/messages/clear_chat.rs]: The test I have has an optional name in the data segment.
https://github.com/museun/twitchchat/blob/master/src/messages/tests.rs#L161-L189

And the docs show the name:
https://dev.twitch.tv/docs/irc/commands#clearchat-twitch-commands 
and
https://dev.twitch.tv/docs/irc/tags#clearchat-twitch-tags

File: src/ng/messages/cap.rs (hunk 2 above)
Comment (museun) [reply src/ng/messages/cap.rs]: As this test shows https://github.com/museun/twitchchat/blob/f6cba76c12e00bc0d5291b7c55e02d1629df34db/src/ng/messages/cap.rs#L73-L81

File: src/ng/messages/clear_chat.rs (hunk 4 above)
Comment (Emilgardis) [reply src/ng/messages/clear_chat.rs]: Ah, my bad!

File: src/ng/messages/cap.rs (hunk 2 above)
Comment (museun) [reply src/ng/messages/cap.rs]: I added the enum in bfe7a2e

File: src/ng/messages/clear_chat.rs (hunk 3 above)
Comment (museun) [reply src/ng/messages/clear_chat.rs]: I don't see that in the docs: 
https://dev.twitch.tv/docs/irc/tags#clearchat-twitch-tags

Do you have a sample for a unit test?

File: src/ng/messages/clear_msg.rs (hunk 5 above)
Comment (museun) [reply src/ng/messages/clear_msg.rs]: Added in 2d3ce81

File: src/ng/messages/clear_msg.rs (hunk 6 above)
Comment (museun) [reply src/ng/messages/clear_msg.rs]: Added in 2d3ce81

File: src/ng/messages/clear_msg.rs (hunk 7 above)
Comment (museun) [reply src/ng/messages/clear_msg.rs]: Added in 2d3ce81

File: src/ng/messages/privmsg.rs (hunk 9 above)
Comment (museun) [reply src/ng/messages/privmsg.rs]: Added in d2aee9d

File: src/ng/messages/user_state.rs (hunk 10 above)
Comment (museun) [reply src/ng/messages/user_state.rs]: I also don't see that in https://dev.twitch.tv/docs/irc/tags#userstate-twitch-tags

Do you have a sample for a test?

File: src/ng/messages/whisper.rs (hunk 1 above)
Comment (museun) [reply src/ng/messages/whisper.rs]: Added the missing fields.

The serde question can be reviewed later.

**Comment by museun** [comment]
I think I'm ready to merge this into master.

I still won't be ready for a release. But this design should be at least on par with features of the previous release.

I have the readme gutted and disabled all of the examples. 
I'll have to go through and all back all of the doc tests/examples.

The current design shouldn't need much updating from users.

Some highlights
* Messages types do not have public fields. use the same-named getters instead.
* Messages will only ever have at most 1 allocation.
* All of the messages allow you to get the 'raw' (original) message out of them.
* `AsOwned`/`Reborrow`/`Event`/`EventMapped`/etc traits are gone. There is just one now: `FromIrcMessage`
* The IRC message parsing is a lot simpler because of the afformentioned trait removal.
* The connector idea has been expanded and will allow for custom domains/addresses and future addition of any new runtime. Currently `smol`, `async_std`, `tokio`, `async_io` are supported.
* The 'async' portion of the crate is runtime agnostic. No more tokio!
* The dispatcher can be used outside of this crate. There is also an async version.
* There is now `AsyncDecoder`/`AsyncEncoder`/`Decoder`/`Encoder` which can be used outside of this crate.
* The writer is now just an `Encodable` trait which can be implemented on your types.
* Because of this change, a bunch of types/functions have been added to create these Encodable types.
* Strings/bytes are also encodable.
* The rate limiter can both be used in sync and async contexts.
* Message parsing can be done outside of this crate (the new `Validator` trait).

**Pull request merged**